contracts/dual_token/DualStakingToken.sol

```diff
@@ -0,0 +1,333 @@
+    address public owner;
+
+    // Mapping of staker account address => Staker info struct
+    mapping(uint256 => StakerInfo) public mapStakerStakerInfos;
+    /// Mapping of staked OLAS service multisigs
+    mapping(address => bool) public mapMutisigs;
+
+    /// @dev DualStakingToken constructor.
+    /// @param _serviceRegistry Service registry address.
+    /// @param _stakingToken Staking token address.
+    /// @param _stakingTokenAmount Staking token amount.
+    /// @param _rewardRatio Staking token ratio to OLAS rewards in 1e18 form.
+    constructor(address _serviceRegistry, address _stakingToken, uint256 _stakingTokenAmount, uint256 _rewardRatio) {
+        // Check for zero addresses
+        if (_serviceRegistry == address(0) || _stakingToken == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Check for zero values
+        if (_stakingTokenAmount == 0 || _rewardRatio == 0) {
+            revert ZeroValue();
+        }
+
+        serviceRegistry = _serviceRegistry;
+        stakingToken = _stakingToken;
+
+        stakingTokenAmount = _stakingTokenAmount;
+        rewardRatio = _rewardRatio;
+
+        owner = msg.sender;
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @notice The balance is always greater or equal the amount, as follows from the Base contract logic.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal {
+        // Update the contract balance
+        balance -= amount;
+
+        SafeTransferLib.safeTransfer(stakingToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Changes contract owner address.
+    /// @param newOwner Address of a new owner.
+    function changeOwner(address newOwner) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for the zero address
+        if (newOwner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        owner = newOwner;
+        emit OwnerUpdated(newOwner);
+    }
+
+    function changeStakingTokenParams(uint256 newStakingTokenAmount, uint256 newRewardRatio) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for zero values
+        if (newStakingTokenAmount == 0 || newRewardRatio == 0) {
+            revert ZeroValue();
+        }
+
+        stakingTokenAmount = newStakingTokenAmount;
+        rewardRatio = newRewardRatio;
+
+        emit StakingTokenParamsUpdated(newStakingTokenAmount, newRewardRatio);
+    }
+
+    /// @dev Stakes OLAS service Id and required staking token amount.
+    /// @param serviceId OLAS driven service Id.
+    /// @param stakingInstance Staking instance address.
+    function stake(uint256 serviceId, address stakingInstance) external {
```

Comment:
IMO DualStakingToken should be scoped to a single staking instance. Otherwise we create a pooled system here whereas on the staking instance level nothing is pooled across instances.

Reply:
Ok, we can do that, but via setters, since all the contracts are dependent on each other:
stakingInstance (can't be deployed without activityChecker) -> activityChecker (can't be deployed without dualStakingToken) -> dualStakingToken -> stakingInstance.

Alternative way - `activityChecker` has a temporary ownership to set the `dualStakingToken` address when it's deployed.

Reply:
I agree with David, this would also simplify the FE work

Reply:
OK due to deployment we can have temporary ownership but this needs be zeroed then

Reply:
Yep, that is already in place